scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java

```diff
@@ -1799,7 +1802,39 @@ protected void processDatagram(DatagramPacket packet, InetSocketAddress router)
 		}
 
 		final Record firstRecord = records.get(0);
+		final ConnectionId connectionId = firstRecord.getConnectionId();
 
+		if (beforeConnectionRetrievalAction == null) {
+			processRecords(records, peerAddress, connectionId, router);
+		} else {
+			final Thread usedReceiverThread = Thread.currentThread();
+			final Runnable processingPart = new Runnable() {
+				@Override
+				public void run() {
+					processRecords(records, peerAddress, connectionId, router);
```

Comment:
There is no need to execute the `processRecords` from the receiver thread or executor.
`executeInbound` will execute the job anyway using the executor. The code executed by the calling thread of `processRecords` is very small and simple.